common/src/main/java/net/william278/husksync/database/MongoDatabase.java

```diff
@@ -0,0 +1,362 @@
+        return CompletableFuture.supplyAsync(() -> {
+            DeleteResult deleteResult = collectionUsersData.deleteOne(
+                    and(
+                            eq("player_uuid", user.uuid.toString()),
+                            eq("version_uuid", versionUuid.toString())
+                    )
+            );
+            return deleteResult.getDeletedCount() > 0;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> setUserData(@NotNull User user, @NotNull UserData userData,
+                                               @NotNull DataSaveCause saveCause) {
+        return CompletableFuture.runAsync(() -> {
+            final DataSaveEvent dataSaveEvent = (DataSaveEvent) getEventCannon().fireDataSaveEvent(user,
+                    userData, saveCause).join();
+            if (!dataSaveEvent.isCancelled()) {
+                final UserData finalData = dataSaveEvent.getUserData();
+                collectionUsersData.insertOne(
+                        new Document("player_uuid", user.uuid.toString())
+                                .append("version_uuid", UUID.randomUUID().toString())
+                                .append("timestamp", Timestamp.from(Instant.now()).toString())
+                                .append("save_cause", saveCause.name())
+                                .append("data", new Binary(getDataAdapter().toBytes(finalData)))
+                );
+                // The Binary class here is the equivalent of blob in SQL.
+            }
+        }).thenRun(() -> rotateUserData(user).join());
+    }
+
+    @Override
+    public CompletableFuture<Void> pinUserData(@NotNull User user, @NotNull UUID versionUuid) {
+        return CompletableFuture.runAsync(() ->
+                        collectionUsersData.updateOne(
+                                eq("player_uuid", user.uuid.toString()),
+                                new Document("$set", new Document("pinned", true))
+                        )
+                // The $set here mean it will update the document if it exists with the new value
+        );
+    }
+
+    @Override
+    public CompletableFuture<Void> unpinUserData(@NotNull User user, @NotNull UUID versionUuid) {
+        return CompletableFuture.runAsync(() ->
+                        collectionUsersData.updateOne(
+                                eq("player_uuid", user.uuid.toString()),
+                                new Document("$set", new Document("pinned", false))
+                        )
+                // The $set here mean it will update the document if it exists with the new value
+        );
+    }
+
+    @Override
+    public CompletableFuture<Void> wipeDatabase() {
+        return CompletableFuture.runAsync(() -> collectionUsersData.drop());
+    }
+
+    @Override
+    public void close() {
+        if (mongoClient != null) {
+            mongoClient.close();
+        }
+
+    }
+
+    private User mapUserFromDocument(Document document) {
```

Comment:
Can we do some `@NotNull` annotations on these methods where applicable -- or return `Optional<>` where it can return `null`? Not super important, but it's a good practice to follow for data integrity throughout the plugin